light-client/src/supervisor.rs

```diff
@@ -0,0 +1,302 @@
+                                    Fork::Forked(block) => {
+                                        self.report_evidence(&block);
+                                        forked.push(block.provider);
+                                    }
+                                    Fork::Faulty(block) => {
+                                        self.peers.remove_secondary(&block.provider);
+                                    }
+                                }
+                            }
+
+                            if !forked.is_empty() {
+                                // Fork detected, exiting
+                                bail!(ErrorKind::ForkDetected(forked))
+                            }
+                        }
+                        None => {
+                            // No fork detected, exiting
+                            // TODO: Send to relayer, maybe the run method does this?
+                            return Ok(light_block);
+                        }
+                    }
+                }
+                // Verification failed
+                Err(_err) => {
+                    // Swap primary, and continue with new primary, if any
+                    self.peers.swap_primary()?;
+                    continue;
+                }
+            }
+        }
+
+        bail!(ErrorKind::NoValidPeerLeft)
+    }
+
+    fn report_evidence(&mut self, _light_block: &LightBlock) {
+        ()
+    }
+
+    #[pre(self.peers.primary().is_some())]
+    fn detect_forks(
+        &mut self,
+        light_block: &LightBlock,
+        trusted_state: &LightBlock,
+    ) -> Option<Vec<Fork>> {
+        if self.peers.secondaries().is_empty() {
+            return None;
+        }
+
+        let fork_detector = ProdForkDetector::new(); // TODO: Should be injectable
+        let result =
+            fork_detector.detect_forks(light_block, &trusted_state, self.peers.secondaries());
+
+        match result {
+            ForkDetection::Detected(forks) => Some(forks),
+            ForkDetection::NotDetected => None,
+        }
+    }
+
+    pub fn handler(&mut self) -> Handler {
+        Handler::new(self.sender.clone())
+    }
+
+    // Consume the instance here but return a runtime which will allow interaction
+    // Maybe return an output channnel here?
+    pub fn run(mut self) {
+        loop {
+            let event = self.receiver.recv().unwrap();
+            match event {
+                Event::Terminate(callback) => {
+                    callback.call(());
+                    return;
+                }
+                Event::VerifyToTarget(height, callback) => {
+                    let outcome = self.verify_to_target(height);
+                    callback.call(outcome);
+                }
+                _ => {
+                    // NoOp?
+                }
+            }
+        }
+    }
+}
+
+pub struct Handler {
+    sender: channel::Sender<Event>,
+}
+
+// Assume single handler
```

Comment:
Does that mean, there will only one handler instance?

Reply:
This is a comment from the sketch that should probably be removed 😅